src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/filters/collection/InCollectionFilter.java

```diff
@@ -0,0 +1,177 @@
+  //        return DBFilterBase.getJsonNode(nodeFactory, arrayValue);
+  //    }
+  //
+  //    @Override
+  //    public boolean canAddField() {
+  //        return false;
+  //    }
+
+  public enum Operator {
+    IN,
+    NIN,
+  }
+
+  public InCollectionFilter(Operator operator, String path, List<Object> arrayValue) {
+    super(path);
+    this.arrayValue = arrayValue;
+    this.operator = operator;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+    InCollectionFilter inFilter = (InCollectionFilter) o;
+    return operator == inFilter.operator && Objects.equals(arrayValue, inFilter.arrayValue);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(arrayValue, operator);
+  }
+
+  @Override
+  public BuiltCondition get() {
+    throw new UnsupportedOperationException("For IN filter we always use getALL() method");
+  }
+
+  public List<BuiltCondition> getAll() {
+    List<Object> values = arrayValue;
+    switch (operator) {
+      case IN:
+        if (values.isEmpty()) return List.of();
+        final ArrayList<BuiltCondition> inResult = new ArrayList<>();
+        for (Object value : values) {
+          if (value instanceof Map) {
+            // array element is sub_doc
+            this.indexUsage.textIndexTag = true;
+            inResult.add(
+                BuiltCondition.of(
+                    ConditionLHS.mapAccess("query_text_values", this.getPath()),
+                    BuiltConditionPredicate.EQ,
+                    new JsonTerm(this.getPath(), getHash(new DocValueHasher(), value))));
+          } else if (value instanceof List) {
+            // array element is array
+            this.indexUsage.textIndexTag = true;
+            inResult.add(
+                BuiltCondition.of(
+                    ConditionLHS.mapAccess("query_text_values", this.getPath()),
+                    BuiltConditionPredicate.EQ,
+                    new JsonTerm(this.getPath(), getHash(new DocValueHasher(), value))));
+          } else {
+            this.indexUsage.arrayContainsTag = true;
+            inResult.add(
+                BuiltCondition.of(
+                    ConditionLHS.column(DATA_CONTAINS),
+                    BuiltConditionPredicate.CONTAINS,
+                    new JsonTerm(getHashValue(new DocValueHasher(), getPath(), value))));
+          }
+        }
+        return inResult;
+      case NIN:
+        if (values.isEmpty()) return List.of();
+        if (!this.getPath().equals(DOC_ID)) {
+          final ArrayList<BuiltCondition> ninResults = new ArrayList<>();
+          for (Object value : values) {
+            if (value instanceof Map) {
+              // array element is sub_doc
+              this.indexUsage.textIndexTag = true;
+              ninResults.add(
+                  BuiltCondition.of(
+                      ConditionLHS.mapAccess("query_text_values", this.getPath()),
```

Comment:
note: This needs to be document somewhere -- but not here. Logic of what is stored where (wrt hashes of sub-docs) is not obvious, may look very odd if one sees just this implementation logic.

Reply:
Bigger question of documenting the data model, resolving this comment.